packages/cojson/src/jsonValue.ts

```diff
@@ -4,3 +4,20 @@ export type JsonAtom = string | number | boolean | null;
 export type JsonValue = JsonAtom | JsonArray | JsonObject | RawCoID;
 export type JsonArray = JsonValue[] | readonly JsonValue[];
 export type JsonObject = { [key: string]: JsonValue | undefined };
+
+type AtLeastOne<T, U = {[K in keyof T]: Pick<T, K> }> = Partial<T> & U[keyof U];
+type ExcludeEmpty<T> = T extends AtLeastOne<T> ? T : never; 
+
+export type CoJsonValue<T> = JsonAtom | CoJsonObjectWithIndex<T> | CoJsonArray<T> | RawCoID;
+export type CoJsonArray<T> = CoJsonValue<T>[] | readonly CoJsonValue<T>[];
+
+/**
+ * Since we are forcing Typescript to elaborate the indexes from the given type passing
+ * non-object values to CoJsonObjectWithIndex will return empty o
+ * E.g. 
+ *   CoJsonObjectWithIndex<() => void> --> {}
+ *   CoJsonObjectWithIndex<RegExp> --> {}
+ * 
+ * Applying the ExcludeEmpty type here to make sure we don't accept functions or non-serializable values
+ */
+export type CoJsonObjectWithIndex<T> = ExcludeEmpty<{ [K in keyof T & string]: JsonValue | undefined }>;
```

Comment:
Tried with `export type CoJsonObjectWithIndex<T> = ExcludeEmpty<{ [K in keyof T]: CoJsonValue<T[K]> | undefined }>;` but Typescript got angry:

<img width="538" alt="Screenshot 2024-10-07 at 16 14 49" src="https://github.com/user-attachments/assets/492295da-456d-4b02-9b92-8ac4f46676b6">